### PR TITLE
Evaluate polynomials at equally spaced grid points

### PR DESCRIPTION
**Pull request by postvakje**
Implement Nuttall's algorithm for efficiently evaluating polynomials at
equally spaced grid points.

### CODE REVIEW

File: sympy/polys/polytools.py [hunk 1]
```diff
@@ -3991,6 +3993,71 @@ def ne(f, g, strict=False):
...
+        Transactions on Acoustics, Speech, and Signal Processing, vol. ASSP-35,
+        no. 10, pp. 1486-1487, October 1987.
+        .. [2] S. C. Datta Roy and Shailey Minocha, "A Note on "Efficient
+        Evaluation of Polynomials and Exponentials of Polynomails for
+        Equispaced Arguments",", IEEE Transactions on Signal Processing,
+        vol. 39, no. 11, pp. 2554-2556, November 1991.
+
+        Examples
+        ========
+        >>> from sympy import Poly
```
Comment (smichr) [inline sympy/polys/polytools.py]: insert a blank line after the `====`

**Comment by smichr** [comment]
Did you compare times with this method vs simply generating the expressions with subs?

**Comment by postvakje** [comment]
Here is a quick test:

from polytools import Poly
from sympy.abc import x
f = Poly(5*x**5+4*x**4+3*x**3+2*x**2+x+1)
For the code:
f.eval_grid(1,1,10000,x)
timeit(10) gives 2.242429741534579
For the code:
[f.subs(x,y) for y in range(1,10001,1)]
timeit(10) gives 8.017799153429223

**Comment by skirpichev** [comment]
> For the code

Why do such insane things?  You could do [f(y) for y in range(1,10001,1)] instead, I don't see too much difference:
```python
In [2]: %timeit f.eval_grid(1,1,10000,x)
1 loops, best of 3: 237 ms per loop
In [3]: %timeit [f(y) for y in xrange(1,10001,1)]
1 loops, best of 3: 316 ms per loop
```

**Comment by postvakje** [comment]
Yes, I am also getting about 20-30% speedup in integer arithmetic.  Not great, but still significant.  The speedup in floating point arithmetic is much larger, but Nuttall's algorithm computes the next term based on the previous term, so errors can accumulate.  
Also, I found that [f.subs(x,y) for y in arange(1.0,1001.0,1.0)] runs much faster than [f(y) for y in arange(1.0,1001.0,1.0)].  I wonder why that is?

**Comment by skirpichev** [comment]
> Yes, I am also getting about 20-30% speedup in integer arithmetic.

I don't see your tests (except for above nonsense), sorry.  And I'm not sure that my tests provide such a proof.

> The speedup in floating point arithmetic is much larger

How, why?

**Comment by postvakje** [comment]
Some more tests with integer arithmetic
```
In [19]: f = Poly(5*x**5+4*x**4+3*x**3+2*x**2+x+1)
In [20]: %timeit f.eval_grid(1,1,1000000,x)
1 loops, best of 3: 22.4 s per loop
In [21]: %timeit [f(y) for y in range(1,1000001,1)]
1 loops, best of 3: 31.9 s per loop
```

tests with floating point:
```
In [37]: f2 = Poly(2.2*x**2+4.3*x+2.2)
In [38]: %timeit f2.eval_grid(1.0,1.0,10000,x)
1 loops, best of 3: 214 ms per loop
In [39]: %timeit [f2(y) for y in arange(1.0,10001.0,1.0)]
1 loops, best of 3: 4.96 s per loop
In [40]: %timeit [f2.subs(x,y) for y in arange(1.0,10001.0,1.0)]
1 loops, best of 3: 925 ms per loop
In [41]: %timeit [f2(1.0+y*1.0) for y in range(10000)]
1 loops, best of 3: 5.05 s per loop
In [42]: %timeit [f2.subs(x,1.0+y*1.0) for y in range(10000)]
1 loops, best of 3: 835 ms per loop
```

Note that in this case f2(y) is significantly slower than f2.subs(x,y)

**Comment by smichr** [comment]
I'm not convinced that we really need this; I am -0.

**Comment by czgdp1807** [comment]
Closing this in reference to above comments. Please feel free to re-open this if needed. Thanks for your contributions.

**Pull request closed without merging**